Annotations: Fixed multi-value variables only partially interpolated

pbaumard opened the pull request:
**What this PR does / why we need it**:

Variable interpolation in annotation tag filters implemented in #10163 is very limited.
It only allows one variable without any prefix.
It does not even allow for the "host:$host" described in https://community.grafana.com/t/is-templating-supported-with-built-in-annotations/4247, despite #10163 description.

This PR allows for host:$host and even more interpolations.

For our use case, it would allow to filter based on several criteria.

This is not our exact use case, but for example it allows to filter for "warning:$host" or "error:$host", or even "$alert_level:$alert_type-$host".

**Which issue(s) this PR fixes**:

Fixes: #17234

**Special notes for your reviewer**:

pbaumard committed:
Annotations: Fixed multi-value variables only partially interpolated in templated annotations queries (#17234)

torkelo commented:
Thanks for contributing!

Not sure I agree with all the changes in the templateSrv think this should be handled in the datasource, templateSrv should only handle interpolation of strings. replaceWithoutFormatting seems out of place.

pbaumard commented:
> Not sure I agree with all the changes in the templateSrv think this should be handled in the datasource, templateSrv should only handle interpolation of strings. replaceWithoutFormatting seems out of place.

Can you give any idea on how the logic could be handled in the datasource? Except a copy/paste of replaceWithoutFormatting in datasource, which is some horrible code duplication, I don't see how it could be done.

Returning some object instead of string in templateSrv has been requested by datasource developers for a while (#5532, #5903). I am also developing a datasource plugin and I am using some ugly hack parsing templateSrv regexp output, similar to the current __delimiter__ hack in datasource.ts. Many datasource developers are probably using some other hacks because of templateSrv limitations. A string is not sufficient for all datasource queries, some queries just need a cartesian product such as the annotation query, or queries of custom datasource using some backend with limited features.

pbaumard committed:
Bug fixes

pbaumard committed:
Added list format

pbaumard commented:
@torkelo I pushed a new version with a new "list" format. Does it seem "out of place" for you?

pbaumard committed:
Resolved merge conflict

pbaumard commented:
@marcusolsson Could this PR be reviewed again? I am ready to modify the PR according to a new design given by Grafana team.

pbaumard committed:
Merged master

ryantxu commented on public/app/features/templating/template_srv.ts:
```diff
@@ -258,46 +264,85 @@ export class TemplateSrv {
   }
 
-  replace(target: string, scopedVars?: ScopedVars, format?: string | Function): any {
+  replace(target: string, scopedVars: ScopedVars, format: 'list'): string[];
+  replace(target: string, scopedVars?: ScopedVars, format?: string | Function): string;
+  replace(target: string, scopedVars?: ScopedVars, format?: string | Function): string | string[] {
```
what is this syntax?  (I have not seen it before... what does it do?)  Is it the same as:
```
replace(target: string, scopedVars?: ScopedVars, format?: 'list' | string | Function): string | string[] {
```

pbaumard replied:
It is using [TypeScript Function Overloading](https://www.typescriptlang.org/docs/handbook/functions.html#overloads). This is more precise than 

     replace(target: string, scopedVars?: ScopedVars, format?: 'list' | string | Function): string | string[] {

since it specifically ensures it returns a string array only if format is "list".

torkelo replied:
this feels like it should be another function and not just a format, format just change interpolation

pbaumard replied:
@torkelo What should be the name of the new function? Just `list` as in

`list(target: string, scopedVars: ScopedVars): string[];`

or  maybe `allValues`?

pbaumard committed:
Merge branch 'master' into annotation_tags_full_interpolation

pbaumard committed:
Fixed tests

pbaumard committed:
Master merge

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Merged master

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master

pbaumard committed:
Merge master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch 'master' into annotation_tags_full_interpolation

pbaumard committed:
Merge  master > annotation_tags_full_interpolation

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

torkelo reviewed (commented):
I do see the need for this but just wish there was a simpler / less complex way

pbaumard committed:
Merge master into annotation_tags_full_interpolation

pbaumard committed:
Merge master into annotation_tags_full_interpolation

torkelo commented on public/app/features/templating/template_srv.ts:
```diff
@@ -312,55 +318,95 @@ export class TemplateSrv implements BaseTemplateSrv {
+          })
+        ).map(t => t.join(''));
+      default:
+        return this.replaceWithoutFormatting(target, scopedVars, format)
+          .map(t => this.formatValue(t.value, t.format, t.variable))
+          .join('');
```
not sure what is going on here, how can this work? formatValue just formats the value so in a expression like `my string $variable test` you cannot join the result of formatValue

pbaumard replied:
@torkelo [formatValue](https://github.com/grafana/grafana/pull/18812/files/615460f86a444aa1f9503ea9eb670cb488660da9#diff-ecab3b74c74dc34a9599591b1ab335a5R155) returns the formatted value, so I don't see why map would not work.

In the case "string $variable test", if $variable has values ["var1", "var2"] then the TargetToken objects returned by `replaceWithoutFormatting` would be:

1. {"value":"string "}
2. {"value":["var1","var2"]}
2. {"value":" test"}

This part is tested in [template_srv.test.ts](https://github.com/grafana/grafana/pull/18812/files/615460f86a444aa1f9503ea9eb670cb488660da9#diff-6819db476648351f2780cfdefec09421R293)

By the way @hugohaggmark, why was template_srv.test.ts removed in 00a9af00fc8a4c6d75dedd164da545e80c79a791?

hugohaggmark replied:
@pbaumard probably because many of those tests needed to be rewritten because we're using Redux state now. Thanks for finding that, I'll look into this asap.

hugohaggmark replied:
seems Torkel already added it back, so all is well again in the kingdom of variables :)

torkelo reviewed (commented):
I do see the need for this but struggle to understand the implementation and it feels like it has to be a simpler way

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Added formatValue types

aknuds1 commented:
Looks as if you need to merge with master and resolve conflicts.

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Fix templateSrv.variableExists test to be false instead of null

pbaumard committed:
Merge master into annotation_tags_full_interpolation

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Replace invalid date by empty string instead of null

For consistency and a simple replace function signature

pbaumard committed:
Fixed e2e test empty string format

pbaumard committed:
Fix e2e test empty string format

torkelo reviewed (commented):
Sorry this still feels like it's adding a case that is so different from all the other formats that is makes the code so much more complex and harder to understand. 

Need to search for a cleaner way to handle this. A seperate replaceToList function maybe?

pbaumard committed:
Change 'list' format to replaceToList function

pbaumard committed:
Add scoped vars to replaceToList test

pbaumard committed:
Move replaceToList test after all formats tests

pbaumard committed:
Fix datasource

pbaumard commented:
> Sorry this still feels like it's adding a case that is so different from all the other formats that is makes the code so much more complex and harder to understand.
> 
> Need to search for a cleaner way to handle this. A seperate replaceToList function maybe?

I removed the 'list' format and changed it for a replaceToList function.

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Merge remote branch master into annotation_tags_full_interpolation

pbaumard committed:
Simplify replaceToList

pbaumard committed:
Simplify replaceWithoutFormatting

pbaumard committed:
Improve naming and comments

pbaumard committed:
More readable cartesian product

pbaumard commented:
> Sorry this still feels like it's adding a case that is so different from all the other formats that it makes the code so much more complex and harder to understand.

@torkelo It adds some complexity, but  when I read the code diff, I don't see too much complexity, considering the very powerful feature added. Note that this PR also lowers complexity in datasource.ts code by removing the weird delimiter logic.

Is this opinion about complexity shared by other developers, e.g. by reviewers @ryantxu, @hugohaggmark or @besartberisha?
I tried to simplify the logic in the last commits and added comments. Can you see anything that can make this code simpler to understand? Do you have better ideas for functions or variables naming?

Note: the cartesianProduct method could be replaced by [d3.cross](https://github.com/d3/d3-array#cross) if Grafana upgrades its d3 version

hugohaggmark commented on public/app/features/templating/template_srv.ts:
```diff
@@ -239,56 +246,93 @@ export class TemplateSrv implements BaseTemplateSrv {
-      const variable = this.getVariableAtIndex(variableName);
-      const fmt = fmt2 || fmt3 || format;
+  private static cartesianProduct(arr: any[][]) {
+    return arr.reduce((acc, curr) => acc.flatMap(a => curr.map(c => a.concat([c]))), [[]]);
```
This is a mouthful indeed :)

pbaumard replied:
Yes it is. But it's one of the most readable solutions of https://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
I added a comment for the function.

hugohaggmark commented:
I agree with Torkel that this is a complex solution but I like the simplification in the datasource.

pbaumard committed:
Add cartesianProduct comment and improve var names

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Fix missing replaceToList in VariableQueryEditor test

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Merge master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Fix replaceToList test

pbaumard committed:
Merge remote-tracking branch 'grafana/master' into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Fix prettier error

pbaumard committed:
Fix prettier error with proper prettier version

pbaumard committed:
Merge master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

hugohaggmark reviewed (commented):
@pbaumard thank you for your patience, but we still haven't decided if this is a good way forward.

pbaumard committed:
Merge branch master into annotation_tags_full_interpolation

pbaumard committed:
No need to improve typing for this PR

torkelo commented:
Hi, Thanks for sticking with this change for so long. We have been unable to find a solution that does not add a ton of complexity. So right now this change feels like it's not worth the added complexity & maintenance cost. But all hope is not lost, if more use cases emerge that require this change or if a future refactoring of TemplateSrv what replaces the regex replace with a tokenizer and parser (something that we want to test), that might make this easier to support.
